### ISSUE TITLE
LTI mass update via Mojo::IOLoop causes server response issues

### ISSUE THREAD

**Issue by somiaj** [openwebwork/webwork2#1841]
I have been testing out LTI mass update on a full class with multiple assignments with mojolicious and noticed the following:

+ When using the webwork2-morbo development server, the whole server stops during the update and cannot handle any requests until the update has finished.

+ When using the webwork2.service file which launches via hypnotoad, the very next request after the mass_update is started waits for the update to finish to respond. Though if I send another request it goes through just fine. So it mostly works with hypnotoad, except for that first request after the update has started.

I have tried looking into Mojo::IOLoop to see what is going on, but not finding any useful information as to why that first request after the mass update starts would be waiting for the update to finish, yet any requests after that seem to process just fine.

**Comment by drgrice1** [comment]
With morbo that will happen.  Morbo only runs one worker, and so latter requests have to wait until the first worker finishes.  That is why it is for development only.

With hypnotoad, that should not happen.  I will look into it.

**Comment by drgrice1** [comment]
Testing things with a generic Mojolicious app and a timer that just sleeps for 10 seconds, I see that everything is the same with that.

For morbo, as I said, the process is occupied and until the timer completes no further requests can be processed.  That is to be expected for a development server.

For hypnotoad, when the timer starts it seems that the next request after that from the same client tries to use the same worker.  Sending another request seems to move to another worker, and all is good for that request.  Note that a request from another browser seems to go to another worker anyway and is fine.

In this generic app, I also tested forking the process inside the timer (using Mojo::IOLoop::Subprocess), and that fixes the problem both for morbo and hypnotoad.  Both are able to continue to serve requests even with the same worker.  However, this does not seem to be an option for webwork2 and the LTI mass_update method.  The problem is that the database connection gets interrupted when a new request comes in, and everything explodes.  The mass_update fails with database errors, and the new request dies.  So far worse than the current situation.

For now, we are just going to have to live with how it is I think.  The user that triggers the mass update, will either need to wait, or click on a link a second time.  I suspect that in production with more entropy in the system (more users making requests and such), that this might actually be better as worker processes will be rotated more frequently.

**Comment by taniwallach** [comment]
The current approach to triggering and running a mass-update for LTI essentially hijacks the server thread (when the request is being shut down via `cleanup_register`) when the code decides to actually do an actual mass update. Since the update involves many single LTI transactions to the remote LMS, it is by nature a task which is quite time-consuming. However, there is no real need for the mass-update to run all that quickly or for there to be any response shown "live" in response to the request which triggered it.

I recommend considering whether the "work" of a mass-update can be offloaded to really be run "externally" from the main event loop.

My first thought was whether this could be best handed off to an external script to do, and for the script to be run in the sub-process.

After some reading, it looks like this is something which might be better suitable to be handled via a task submitted to a Minion job queue. See: https://docs.mojolicious.org/Minion and https://docs.mojolicious.org/Mojolicious/Plugin/Minion.

Then all the time-consuming processing would be then be totally offloaded from the regular web-server. Log reports of success/errors could be logged to a new log file for the relevant course. The "external" code (Minion task or script) doing the actual mass-update should open and close its own connection to the database (or at least use a connection not being used by one of the regular workers).

The decision of when an actual mass_update should be triggered could just be made either directly in the `go` method of `lib/WeBWorK/ContentGenerator.pm` or by just running the initial portion of the current `mass_update` subroutine.

---

Sending emails in `lib/WeBWorK/ContentGenerator/Instructor/SendMail.pm` could also probably benefit from being split off to be handled as an "external" task and not kept inside the main process.

---

For future thought - the entire mass_update code seems to exist to try to overcome the problem of failed grade pass-backs. At present, each mass-update will send the LMS all the scores even if they have not been changed after the server was able to last successfully send grades to the LMS. If the database recorded a boolean LMS_was_updated for at the homework set / user level and set it to true after a successfully grade passback to the LMS (possibly confirming with the LMS that the score can be retrieved) and cleared that field whenever a grade change was saved, then far fewer scores would actually need to be sent to the LMS when a mass-update was triggered. We could provide a mechanism to allow staff to force a "full mass-update" either by making it possible to clear the LMS_was_updated field course-wide or provide an course management link to trigger a mass-update which would ignore the value of LMS_was_updated

**Comment by somiaj** [comment]
> For future thought - the entire mass_update code seems to exist to try to overcome the problem of failed grade pass-backs. At present, each mass-update will send the LMS all the scores even if they have not been changed after the server was able to last successfully send grades to the LMS. If the database recorded a boolean LMS_was_updated for at the homework set / user level and set it to true after a successfully grade passback to the LMS (possibly confirming with the LMS that the score can be retrieved) and cleared that field whenever a grade change was saved, then far fewer scores would actually need to be sent to the LMS when a mass-update was triggered. We could provide a mechanism to allow staff to force a "full mass-update" either by making it possible to clear the LMS_was_updated field course-wide or provide an course management link to trigger a mass-update which would ignore the value of LMS_was_updated

I am currently working on a page that an instructor can trigger a mass update of all grades, or just a selected user and/or set (I often find after modifying an assignment or manually grading one, to want to just update a single set vs all of them). In working with this is how I originally ran across this issue. I'll make a draft pull request soon so others can look at it.

Though it would be nice to have some system that can just be triggered, and also one that can be monitored to if it is running or not, this way new updates can't get triggered until the current one has finished.

**Comment by drgrice1** [comment]
I am familiar with Minion, and have already considered it. Unfortunately, Minion requires postgres.  That is the only implemented backend at this point. That is something that webwork2 can not depend on.

**Comment by taniwallach** [comment]
@drgrice1 Sorry - I'm not familiar with it, just found it while trying to consider a better approach to how mass updates can be handled. However, I did recall seeing mention of a SQLite backend in a blog post at https://thecurioustechnoid.com/mojolicious-minion-a-high-performance-job-queue/ and Google found the module https://metacpan.org/pod/Minion::Backend::SQLite .

**Comment by drgrice1** [comment]
Odd, now that I do a Google search for "Minion backends" I find that as well as https://metacpan.org/pod/Minion::Backend::mysql.  The searches I have done before have all had "Mojolicious" in them, and I guess that was to narrow.

I any case, with one of those backends, we can probably use Minion.  Note, that this will add some additional setup though.  Minion is run as a separate app from the Mojolicious app, and so another systemd service file will be needed to control that (or similar such mechanism).  This could perhaps be optional for those that need the functionality provided.  When I get a chance, I will implement it.

**Comment by taniwallach** [comment]
I have been thinking more about the underlying problem, and not just what we can do to fix LTI mass update.

If I understand things correctly after looking at https://docs.mojolicious.org/Mojo/Server/Hypnotoad it seems that the issue with LTI mass_update is that it is a blocking operation. Since Mojolicious/hypnotoad allow a worker to process several requests concurrently (the `clients` setting), when the code of one request gets into a long blocking operation - the worker is just "stuck" from the perspective of the request being processed as well as any later requests allocated to the same worker until the relevant section of blocking code completes.

AFAIK This is quite different from the manner in which requests are assigned to Apache worker. AFAIK in Apache, the workers would not have accepted a new request for processing until the prior one had been completed, which is why in the past we did not need to worry about latency for additional requests being caused by a small number of requests with long processing times, so long as other workers were available reasonably quickly.

If this conclusion is correct, any value of the `clients` setting over 1 will potentially allow "new requests" to be assigned to a worker which is caught up in a "long" blocking operation, which may create very noticeable lags in the responsiveness of the system even under moderate load, as some requests will end up assigned to workers busy on a long blocking request.

Given the fact that essentially all the webwork codebase was written under the assumption that the "worker" running the code was **not** going to be attempting to processes multiple requests concurrently, I don't think we can quickly redesign the WeBWorK code to be more supportive of the potential for concurrent handling of requests.  As such, it would seem that WeBWorK as a `hypnotoad` app should probably be configured to use a very low value of `clients` (possible just 1) and a higher value of `workers` to provide a mechanism to handle multiple requests in parallel.

However, it does seem that we should try to locate the most likely places for significant delays due to inherently slow operations (such as the network operations in LTI grade-passback) and try to refactor such parts of the codebase to either run the code as a background task (if the task is really slow) or to use asynchronous coding practices to allow concurrent request processing while the slow request is being processed.

My impression is still that LTI mass update is probably is case where handing off the work to a background task is the better solution.

---

I think the idea of using Minion for "slow tasks" could probably be of great benefit to some of the other "slow" things, as they could also trigger similar latency on additional requests assigned to the same worker.

@drgrice1 Your comments in `conf/webwork2.mojolicious.dist.yml` hint that relatively large timeout values were set to try to prevent failures of "long" operations (like assigning a set to all users). I suspect that in most cases, such "long" operations are not really things which truly need to complete immediately, and could be handed off to be processed as background tasks.

Other "slow" parts of the system I can think of:
  - scoring a course,
  - maybe also PDF generation
As I see it, these do not really require an "immediate" response, but there may need to be some mechanism to determine that the task was completed.

Any such slow operations which can be offloaded into external tasks would keep the main server more available to process "regular" requests. Use of high priority queues would hopefully mean that we could be reasonably confident that the more urgent tasks would be run reasonable soon after the task was queued for processing, but allow the main server to quickly complete the processing of the "main" request without keeping a worker tied up while the slow process was running.

Minion supports using "spare" fast-lane high-priority queues (https://dev.to/kraih/high-priority-fast-lane-for-the-minion-job-queue-4711) for backends which implemented the necessary changes, including the SQLite backend (https://github.com/Grinnz/Minion-Backend-SQLite/commit/cc67ef907b3a604adfc0970dd7a976685b085ee8). From what I can tell, the MySQL backend does not seem to support the `min_priority` feature. That might be an significant advantage to using the SQLite backend (unless we want to forking the MySQL backend and add the feature).

---

There is a Minion plugin for Mojolicious. Maybe that means that it does not need to be treated as a fully separate app. See:   https://docs.mojolicious.org/Mojolicious/Plugin/Minion . However, from https://docs.mojolicious.org/Minion/Guide it seems that even with the plugin there is a need to take additional steps to start a Minion worker, so in any case there are certainly additional things to start up.

The ideas in https://github.com/openwebwork/webwork2/pull/1842 could take advantage of the option to set priorities for tasks sent to Minion, so that "smaller" LTI updates would get processed before bigger ones.

---

I'm not sure how using Minion will work with running WeBWorK in a Docker container, but I assume that docker-entrypoint.sh can probably be used to start up the Minion worker in the same container. Even if running multiple things in a single Docker container is not something typical recommended, in this case it seems relatively reasonable to me.

**Comment by drgrice1** [comment]
@taniwallach: A couple of quick notes on your comments.

Scoring and PDF generation do need an immediate response.  So those won't work in an external task as you suggest.  Also, they generally are not that slow.

Assigning a set to all users is not a problem in a general setting where you only have a few hundred users.  It is only a problem when you have a few thousand users.  An extreme case.

**Comment by somiaj** [comment]
This should be fixed with the new job-queue, closing.

**Comment by somiaj** [comment]
Never actually closed this.
